mlir/lib/Dialect/SparseTensor/Transforms/SparseReinterpretMap.cpp

```diff
@@ -73,128 +353,57 @@ static bool hasNonIdentityOperandsOrResults(Operation *op) {
 //===----------------------------------------------------------------------===//
 
 /// Sparse rewriting rule for the generic `linalg` operation.
-struct GenericOpReinterpretMap : public OpRewritePattern<linalg::GenericOp> {
+struct GenericOpReinterpretMap
+    : public DemapInsRewriter<GenericOpReinterpretMap, linalg::GenericOp> {
 public:
-  GenericOpReinterpretMap(MLIRContext *context)
-      : OpRewritePattern<linalg::GenericOp>(context) {}
+  using DemapInsRewriter::DemapInsRewriter;
+  LogicalResult rewriteOp(linalg::GenericOp linalgOp, OpAdaptor adaptor,
+                          PatternRewriter &rewriter) const {
+    // Only rewrite single output operations with pure (sparse) tensor
+    // semantics.
+    if (linalgOp.getNumDpsInits() != 1 || !linalgOp.hasTensorSemantics() ||
+        !hasAnySparseOperandOrResult(linalgOp) ||
+        !hasNonIdentityOperandsOrResults(linalgOp))
+      return failure();
 
-  LogicalResult matchAndRewrite(linalg::GenericOp linalgOp,
-                                PatternRewriter &rewriter) const override {
-    // Only rewrite single output operations with pure tensor semantics.
-    if (linalgOp.getNumDpsInits() != 1 || !linalgOp.hasTensorSemantics())
+    // Try translating the index map.
+    if (failed(translateMap(linalgOp, rewriter)))
       return failure();
-    // Scan all operands, inspect sparse tensors.
-    //
-    // TODO: generalize this proof-of-concept algorithm, since the current
-    //       implementation accepts only simple indexing maps, and one
-    //       non-permutation sparse tensor, which must have an identity
-    //       indexing map and be the output.
-    //
-    OpOperand *tx = nullptr;
-    for (OpOperand &t : linalgOp->getOpOperands()) {
-      // Ensure every index map is "simple".
-      const auto map = linalgOp.getMatchingIndexingMap(&t);
-      for (unsigned i = 0, n = map.getNumResults(); i < n; i++)
-        if (map.getResult(i).getKind() != AffineExprKind::DimId)
-          return failure();
-      // Inspect sparse operands.
-      auto stt = tryGetSparseTensorType(t.get());
-      if (stt && stt->hasEncoding()) {
-        if (stt->isPermutation())
-          continue;
-        assert(stt->getDimRank() < stt->getLvlRank()); // only allowed non-perm
-        if (tx)
-          return failure(); // more than one non-perm
-        if (!map.isIdentity())
-          return failure(); // no ID indexing map on the non-perm
-        tx = &t;
-      }
-    }
-    // Found a non-permutation, rewrite when this is the output.
-    if (tx && tx == linalgOp.getDpsInitOperand(0)) {
-      auto stt = getSparseTensorType(tx->get());
-      auto demap = genDemap(rewriter, stt.getEncoding(), tx->get());
-      auto newOp = genGenericLinalg(rewriter, linalgOp, stt, demap);
-      auto remap = genRemap(rewriter, stt.getEncoding(), newOp.getResult(0));
-      rewriter.replaceOp(linalgOp, remap);
-      return success();
+
+    // Must only have one result
```

Comment:
period in the end.

Reply:
A "must" comment is never very useful.Is it assumed, or it is enforced.
I would stick with descriptive comment, like

// On success, update the linalg op fields, including the single result, in place.